field_audit/models.py

```diff
@@ -0,0 +1,259 @@
+        )
+
+    # "no branch" note: tests only run on postgres _or_ sqlite, never both
+    if check_engine_sqlite():  # pragma: no branch
+        _by_type_and_username = by_type_and_username
+
+        def by_type_and_username(self, user_type, username):
+            """Support SQLite (for development) and Oracle (because it comes
+            along for free).
+
+            If these DB flavors are not needed at import time, this "extra db"
+            support is never defined.
+            """
+            if check_engine_sqlite(settings.DATABASES[self.db]["ENGINE"]):
+                # Oracle and SQLite do not support `contains` queries
+                # see: https://docs.djangoproject.com/en/4.0/topics/db/queries/#std:fieldlookup-jsonfield.contains  # noqa: E501
+                return self.filter(
+                    changed_by__user_type=user_type,
+                    changed_by__username=username,
+                )
+            # "no cover" note: tests only run on postgres _or_ sqlite, never
+            # both
+            return self._by_type_and_username(user_type, username)  # pragma: no cover  # noqa: E501
+
+
+class DefaultAuditEventManager(AuditEventManager):
+    """Default Manager for the AuditEvent model. Contains convenience methods
+    for the default auditors, which may not be desirable to subclass if
+    downstream projects wish to define custom auditor chains.
+    """
+
+    def by_system_user(self, username):
+        system_types = [USER_TYPE_TTY, USER_TYPE_PROCESS]
+        return self.filter(
+            changed_by__user_type__in=system_types,
+            changed_by__username=username,
+        )
+
+    def by_tty_user(self, username):
+        return self.by_type_and_username(USER_TYPE_TTY, username)
+
+    def by_process_user(self, username):
+        return self.by_type_and_username(USER_TYPE_PROCESS, username)
+
+    def by_request_user(self, username):
+        return self.by_type_and_username(USER_TYPE_REQUEST, username)
+
+
+def get_manager(attr_suffix, default):
+    """Returns an instantiated manager, possibly one defined in settings.
+
+    If the manager is defined in settings, it must be a subclass of
+    ``django.db.models.Manager``.
+
+    :param attr_suffix: Suffix (appended to ``FIELD_AUDIT_``) of settings
+        attribute for the manager class path.
+    :param default: Manager class. Returned if attribute isn't defined.
+    """
+    settings_attr = f"FIELD_AUDIT_{attr_suffix}"
+    try:
+        class_path = getattr(settings, settings_attr)
+    except AttributeError:
+        return default()
+    desc = f"{settings_attr!r} value"
+    return class_import_helper(class_path, desc, models.Manager)()
+
+
+def get_date():
+    return datetime.utcnow()
+
+
+class AuditEvent(models.Model):
+    event_date = models.DateTimeField(default=get_date, db_index=True)
+    object_class_path = models.CharField(db_index=True, max_length=255)
```

Comment:
if we are storing a class path, what happens if a class get's moved or renamed? Perhaps the decorator could also provide a 'name' for the model that will remain consistent?

Reply:
Added in 25400ab.